fem/fe/fe_pyramid.hpp

```diff
@@ -0,0 +1,489 @@
+
+   /// Pyramid "Affine" Coordinates
+   static real_t lam1(real_t x, real_t y, real_t z)
+   { return CheckZ(z) ? (1.0 - x - z) * (1.0 - y - z) / (1.0 - z): 0.0; }
+   static real_t lam2(real_t x, real_t y, real_t z)
+   { return CheckZ(z) ? x * (1.0 - y - z) / (1.0 - z): 0.0; }
+   static real_t lam3(real_t x, real_t y, real_t z)
+   { return CheckZ(z) ? x * y / (1.0 - z): 0.0; }
+   static real_t lam4(real_t x, real_t y, real_t z)
+   { return CheckZ(z) ? (1.0 - x - z) * y / (1.0 - z): 0.0; }
+   static real_t lam5(real_t x, real_t y, real_t z)
+   { return CheckZ(z) ? z : 1.0; }
+
+   /// Gradients of the "Affine" Coordinates
+   static Vector grad_lam1(real_t x, real_t y, real_t z);
+   static Vector grad_lam2(real_t x, real_t y, real_t z);
+   static Vector grad_lam3(real_t x, real_t y, real_t z);
+   static Vector grad_lam4(real_t x, real_t y, real_t z);
+   static Vector grad_lam5(real_t x, real_t y, real_t z);
+
+   /// Two component vectors associated with edges touching the apex
+   static Vector lam15(real_t x, real_t y, real_t z)
+   { return Vector({lam1(x, y, z), lam5(x, y, z)}); }
+   static Vector lam25(real_t x, real_t y, real_t z)
+   { return Vector({lam2(x, y, z), lam5(x, y, z)}); }
+   static Vector lam35(real_t x, real_t y, real_t z)
+   { return Vector({lam3(x, y, z), lam5(x, y, z)}); }
+   static Vector lam45(real_t x, real_t y, real_t z)
+   { return Vector({lam4(x, y, z), lam5(x, y, z)}); }
+
+   /// Gradients of the above two component vectors
+   static DenseMatrix grad_lam15(real_t x, real_t y, real_t z);
+   static DenseMatrix grad_lam25(real_t x, real_t y, real_t z);
+   static DenseMatrix grad_lam35(real_t x, real_t y, real_t z);
+   static DenseMatrix grad_lam45(real_t x, real_t y, real_t z);
+
+   static Vector lam15_grad_lam15(real_t x, real_t y, real_t z);
+   static Vector lam25_grad_lam25(real_t x, real_t y, real_t z);
+   static Vector lam35_grad_lam35(real_t x, real_t y, real_t z);
+   static Vector lam45_grad_lam45(real_t x, real_t y, real_t z);
+
+   /// Three component vectors associated with triangular faces
+   static Vector lam125(real_t x, real_t y, real_t z)
+   { return Vector({lam1(x, y, z), lam2(x, y, z), lam5(x, y, z)}); }
+   static Vector lam235(real_t x, real_t y, real_t z)
+   { return Vector({lam2(x, y, z), lam3(x, y, z), lam5(x, y, z)}); }
+   static Vector lam345(real_t x, real_t y, real_t z)
+   { return Vector({lam3(x, y, z), lam4(x, y, z), lam5(x, y, z)}); }
+   static Vector lam435(real_t x, real_t y, real_t z)
+   { return Vector({lam4(x, y, z), lam3(x, y, z), lam5(x, y, z)}); }
+   static Vector lam415(real_t x, real_t y, real_t z)
+   { return Vector({lam4(x, y, z), lam1(x, y, z), lam5(x, y, z)}); }
+   static Vector lam145(real_t x, real_t y, real_t z)
+   { return Vector({lam1(x, y, z), lam4(x, y, z), lam5(x, y, z)}); }
+
+   /// Vector functions related to the normals to the triangular faces
+   static Vector lam125_grad_lam125(real_t x, real_t y, real_t z);
```

Comment:
Are these explicit formulae for `s_0 \nabla s_1 \times \nabla s_2 + s_1 \nabla s_2 \times \nabla s_0 + s_2 \nabla s_0 \times \nabla s_1`?

I tried to verify this guess, but I think that I would need Mathematica.

Reply:
I have added a more detailed comment